File: src/test/java/com/pubnub/api/PubNubTest.java
package com.pubnub.api;

import com.pubnub.api.enums.PNReconnectionPolicy;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;

public class PubNubTest {
    private PubNub pubnub;
    private PNConfiguration pnConfiguration;

    @Before
    public void beforeEach() throws IOException, PubNubException {
        pnConfiguration = new PNConfiguration(PubNub.generateUUID());
        pnConfiguration.setSubscribeKey("demo");
        pnConfiguration.setPublishKey("demo");
        pnConfiguration.setUseRandomInitializationVector(false);
    }

    @After
    public void cleanup() {
        pubnub.forceDestroy();
        pubnub = null;
    }

    @Test
    public void testCreateSuccess() {
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals(true, pubnub.getConfiguration().isSecure());
        Assert.assertNotNull("pubnub object is null", pubnub);
        Assert.assertNotNull(pubnub.getConfiguration());
        Assert.assertEquals("https://ps.pndsn.com", pubnub.getBaseUrl());
    }

    @Test
    public void testEncryptCustomKey() throws PubNubException {
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals("iALQtn3PfIXe74CT/wrS7g==", pubnub.encrypt("test1", "cipherKey").trim());

    }

    @Test
    public void testEncryptConfigurationKey() throws PubNubException {
        pnConfiguration.setCipherKey("cipherKey");
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals("iALQtn3PfIXe74CT/wrS7g==", pubnub.encrypt("test1").trim());

    }

    @Test
    public void testDecryptCustomKey() throws PubNubException {
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals("test1", pubnub.decrypt("iALQtn3PfIXe74CT/wrS7g==", "cipherKey").trim());

    }

    @Test
    public void testDecryptConfigurationKey() throws PubNubException {
        pnConfiguration.setCipherKey("cipherKey");
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals("test1", pubnub.decrypt("iALQtn3PfIXe74CT/wrS7g==").trim());

    }

    @Test
    public void testPNConfiguration() {
        pnConfiguration.setSubscribeTimeout(3000);
        pnConfiguration.setConnectTimeout(4000);
        pnConfiguration.setNonSubscribeRequestTimeout(5000);
        pnConfiguration.setReconnectionPolicy(PNReconnectionPolicy.NONE);
        pubnub = new PubNub(pnConfiguration);

        Assert.assertNotNull("pubnub object is null", pubnub);
        Assert.assertNotNull(pubnub.getConfiguration());
        Assert.assertEquals("https://ps.pndsn.com", pubnub.getBaseUrl());
        Assert.assertEquals(3000, pnConfiguration.getSubscribeTimeout());
        Assert.assertEquals(4000, pnConfiguration.getConnectTimeout());
        Assert.assertEquals(5000, pnConfiguration.getNonSubscribeRequestTimeout());
    }

    @Test(expected = PubNubException.class)
    public void testDecryptNull() throws PubNubException {
        pnConfiguration.setCipherKey("cipherKey");
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals("test1", pubnub.decrypt(null).trim());
    }

    @Test(expected = PubNubException.class)
    public void testDecryptNull_B() throws PubNubException {
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals("test1", pubnub.decrypt(null, "cipherKey").trim());
    }

    @Test
    public void getVersionAndTimeStamp() {
        pubnub = new PubNub(pnConfiguration);
        String version = pubnub.getVersion();
        int timeStamp = pubnub.getTimestamp();
        Assert.assertEquals("6.2.0", version);
        Assert.assertTrue(timeStamp > 0);
    }

    @Test(expected = PubNubException.class)
    public void testEcryptNull() throws PubNubException {
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals("test1", pubnub.encrypt(null));
    }

    @Test(expected = PubNubException.class)
    public void testEcryptNull_B() throws PubNubException {
        pubnub = new PubNub(pnConfiguration);
        Assert.assertEquals("test1", pubnub.encrypt(null, "chiperKey"));
    }
}

File: src/test/kotlin/com/pubnub/api/PubNubTest.kt
package com.pubnub.api

import com.pubnub.api.enums.PNReconnectionPolicy
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class PubNubTest : BaseTest() {

    lateinit var config: PNConfiguration

    override fun onBefore() {
        config = PNConfiguration().apply {
            subscribeKey = "demo"
            publishKey = "demo"
        }
    }

    override fun onAfter() {
        pubnub.destroy()
    }

    @Test
    fun testCreateSuccess() {
        pubnub = PubNub(config)
        assertEquals(true, pubnub.configuration.secure)
        assertEquals("https://ps.pndsn.com", pubnub.baseUrl())
    }

    @Test
    @Throws(PubNubException::class)
    fun testEncryptCustomKey() {
        pubnub = PubNub(config)
        assertEquals("iALQtn3PfIXe74CT/wrS7g==", pubnub.encrypt("test1", "cipherKey").trim())
    }

    @Test
    @Throws(PubNubException::class)
    fun testEncryptConfigurationKey() {
        config.cipherKey = "cipherKey"
        pubnub = PubNub(config)
        assertEquals("iALQtn3PfIXe74CT/wrS7g==", pubnub.encrypt("test1").trim())
    }

    @Test
    @Throws(PubNubException::class)
    fun testDecryptCustomKey() {
        pubnub = PubNub(config)
        assertEquals("test1", pubnub.decrypt("iALQtn3PfIXe74CT/wrS7g==", "cipherKey").trim())
    }

    @Test
    @Throws(PubNubException::class)
    fun testDecryptConfigurationKey() {
        config.cipherKey = "cipherKey"
        pubnub = PubNub(config)
        assertEquals("test1", pubnub.decrypt("iALQtn3PfIXe74CT/wrS7g==").trim())
    }

    @Test
    fun testconfig() {
        config.subscribeTimeout = 3000
        config.connectTimeout = 4000
        config.nonSubscribeRequestTimeout = 5000
        config.reconnectionPolicy = PNReconnectionPolicy.NONE
        pubnub = PubNub(config)
        assertEquals("https://ps.pndsn.com", pubnub.baseUrl())
        assertEquals(3000, config.subscribeTimeout)
        assertEquals(4000, config.connectTimeout)
        assertEquals(5000, config.nonSubscribeRequestTimeout)
    }

    @Test
    fun getVersionAndTimeStamp() {
        pubnub = PubNub(config)
        val version = pubnub.version
        val timeStamp = pubnub.timestamp()
        assertEquals("0.0.1-canary", version)
        assertTrue(timeStamp > 0)
    }

}